### PR TITLE
Set up base test case with local SQL DB access

### PR DESCRIPTION
**Pull request by hhdqirui**
Part of #18

TestContainer is integrated. A docker container is able to start up for testing and SQL queries are able to be run inside the container.

For now, tests are added for `EntityDb` and creating courses for `CourseDb`. This PR focuses on the test skeleton and more tests will be added in the future.

The removal of the static block in `HibernateUtil` is to help create a different `sessionFactory` for testing.

### CODE REVIEW

File: src/test/java/teammates/storage/sql/CoursePostgresDbTest.java
Comment (jianhandev) [inline src/test/java/teammates/storage/sql/CoursePostgresDbTest.java]: Do you think it is possible to move this to the `BaseTestCaseWithLocalDatabaseAccess` class, so that we don't need to repeat this code for all the test classes requiring local database access?

Comment (hhdqirui) [reply src/test/java/teammates/storage/sql/CoursePostgresDbTest.java:25]: These few lines of code are just for testing that the test container works. They will be removed when real tests are written.

Comment (hhdqirui) [reply src/test/java/teammates/storage/sql/CoursePostgresDbTest.java]: Yes. For now I will move it to `BaseTestCaseWithLocalPostgresSqlDatabaseAccess` before the `BaseTestCaseWithLocalDatabaseAccess ` is eventually removed.

File: src/test/java/teammates/logic/api/LogicNewExtension.java
Comment (zhaojj2209) [inline src/test/java/teammates/logic/api/LogicNewExtension.java:2]: Would be good to add a javadoc here mirroring that in `LogicExtension.java`.

File: src/test/java/teammates/test/BaseTestCaseWithLocalPostgresSqlDatabaseAccess.java
Comment (zhaojj2209) [inline src/test/java/teammates/test/BaseTestCaseWithLocalPostgresSqlDatabaseAccess.java]: Seems that there's no method being called here to change the session factory back to the local session factory after testing?

File: src/test/java/teammates/storage/sql/CoursePostgresDbTest.java
Comment (zhaojj2209) [inline src/test/java/teammates/storage/sql/CoursePostgresDbTest.java]: The tests are already in the `sql` folder, so the class names can be kept the same as their counterparts in the `api` folder (`CoursesDbTest`) . This reduces the need for us to refactor the class names again down the line. Same applies to `EntitiesDbTest`.

File: src/test/java/teammates/test/BaseTestCaseWithLocalPostgresSqlDatabaseAccess.java
Comment (zhaojj2209) [inline src/test/java/teammates/test/BaseTestCaseWithLocalPostgresSqlDatabaseAccess.java]:
```suggestion
public abstract class BaseTestCaseWithLocalSqlDatabaseAccess extends BaseTestCaseWithSqlDatabaseAccess {
```
It would be good for the class names to be database agnostic.

File: src/test/java/teammates/test/BaseTestCaseWithPostgresSqlDatabaseAccess.java
Comment (zhaojj2209) [inline src/test/java/teammates/test/BaseTestCaseWithPostgresSqlDatabaseAccess.java]:
```suggestion
public abstract class BaseTestCaseWithSqlDatabaseAccess extends BaseTestCase {
```
Same here; also, missing space.

**Review by zhaojj2209 — APPROVED** [review]

**Review by zhaojj2209 — CHANGES_REQUESTED** [review]
Apologies, clicked the wrong option when submitting review. Do address the above comments.

File: src/test/java/teammates/storage/sql/CoursePostgresDbTest.java
Comment (hhdqirui) [reply src/test/java/teammates/storage/sql/CoursePostgresDbTest.java]: Fixed

File: src/test/java/teammates/test/BaseTestCaseWithLocalPostgresSqlDatabaseAccess.java
Comment (hhdqirui) [reply src/test/java/teammates/test/BaseTestCaseWithLocalPostgresSqlDatabaseAccess.java]: We can only create the `sessionFactory` when there is a postgre docker container running. Therefore, if we want to reset it back to the local session factory, when we want to do testing, we would also need to start up the dev postgre container, which I feel it is not the best way to do it.

File: src/test/java/teammates/logic/api/LogicNewExtension.java
Comment (hhdqirui) [reply src/test/java/teammates/logic/api/LogicNewExtension.java:2]: Added

File: src/test/java/teammates/test/BaseTestCaseWithLocalPostgresSqlDatabaseAccess.java
Comment (zhaojj2209) [reply src/test/java/teammates/test/BaseTestCaseWithLocalPostgresSqlDatabaseAccess.java]: Got it, as long as the test container is isolated to the tests and there is no situation where the session factory is still connected to the test container even after the tests are done running.

**Review by zhaojj2209 — APPROVED** [review]

File: src/test/java/teammates/test/BaseTestCaseWithLocalSqlDatabaseAccess.java
Comment (jianhandev) [inline src/test/java/teammates/test/BaseTestCaseWithLocalSqlDatabaseAccess.java]: If I'm not wrong this can be removed.

File: src/test/java/teammates/test/BaseTestCaseWithLocalSqlDatabaseAccess.java
Comment (jianhandev) [inline src/test/java/teammates/test/BaseTestCaseWithLocalSqlDatabaseAccess.java]: Eventually we should run the migration script using Flyway/Liquibase, so that there's no need to sync the migration files in `src/main/resources/db/migration` and `src/test/resources/sql`, but can leave that to another issue

File: src/main/java/teammates/common/util/HibernateUtil.java
Comment (jianhandev) [inline src/main/java/teammates/common/util/HibernateUtil.java]: If possible can consider removing the duplication with the configuration in `getSessionFactory` since the only three different properties are `username`, `password` and `url`.

File: src/test/java/teammates/test/BaseTestCaseWithLocalSqlDatabaseAccess.java
Comment (jianhandev) [reply src/test/java/teammates/test/BaseTestCaseWithLocalSqlDatabaseAccess.java]: After #32 is merged can call `getFlywayInst(...).migrate()` to run the migrations.

File: src/test/java/teammates/test/BaseTestCaseWithLocalSqlDatabaseAccess.java
Comment (hhdqirui) [reply src/test/java/teammates/test/BaseTestCaseWithLocalSqlDatabaseAccess.java]: Yes, this was for testing container only. It is removed.

File: src/main/java/teammates/common/util/HibernateUtil.java
Comment (hhdqirui) [reply src/main/java/teammates/common/util/HibernateUtil.java]: Ok! Code refactored

**Review by jianhandev — APPROVED** [review]
LGTM. I have merged #32 into this branch as well.

**Pull request merged**
